sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -364,83 +366,263 @@ struct codec_dai dai_lists[] = {
+		} else {
+			name = devm_kasprintf(dev, GFP_KERNEL,
+					      "SDW%d-Capture", link_id);
+			dai_links[i].dpcm_capture = 1;
+		}
+
+		if (!name)
+			return -ENOMEM;
+
+		cpu_name = devm_kasprintf(dev, GFP_KERNEL,
+					  "SDW%d Pin%d", link_id, j++);
+		if (!cpu_name)
+			return -ENOMEM;
+
+		init_dai_link(dai_links, i, name, cpus, cpu_name,
+			      codec, link->num_adr);
+	}
+
+	if (part_id == 0x711)
+		dai_links[id].init = headset_init;
+
+	if ((link_id == 0x2 || link->num_adr > 1) && part_id == 0x1308)
+		dai_links[id].init = second_spk_init;
+
+	return dai_num;
+}
+
+static int sof_card_dai_links_create(struct device *dev,
+				     struct snd_soc_acpi_mach *mach,
+				     struct snd_soc_card *card)
+{
+	struct snd_soc_dai_link_component *idisp_components;
+	struct snd_soc_acpi_mach_params *mach_params;
+	int ssp_num = 0, sdw_num, hdmi_num = 0;
+	struct snd_soc_dai_link_component *cpus;
+	const struct snd_soc_acpi_link_adr *link;
+	struct snd_soc_dai_link *links;
+	const char *board;
+	int num_links;
+	int i, id = 0;
+	int val;
 
 #if IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)
-	{
-		.name = "iDisp1",
-		.id = 5,
-		.init = hdmi_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		SND_SOC_DAILINK_REG(idisp1_pin, idisp1_codec, platform),
-	},
-	{
-		.name = "iDisp2",
-		.id = 6,
-		.init = hdmi_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		SND_SOC_DAILINK_REG(idisp2_pin, idisp2_codec, platform),
-	},
-	{
-		.name = "iDisp3",
-		.id = 7,
-		.init = hdmi_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		SND_SOC_DAILINK_REG(idisp3_pin, idisp3_codec, platform),
-	},
+	board = dmi_get_system_info(DMI_BOARD_NAME);
+	if (strstr(board, "TigerLake"))
+		hdmi_num = 4;
+	else
+		hdmi_num = 3;
 #endif
-	{
-		.name = "SDW2-Playback",
-		.id = 3,
-		.init = second_spk_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw2_pin2, sdw2_codec, platform),
-	},
-};
+
+	if (snd_soc_acpi_find_machine(mach))
+		ssp_num = 1;
+
+	mach_params = &mach->mach_params;
+	sdw_num = get_sdw_dai_link_num(mach_params);
+
+	dev_dbg(dev, "sdw %d, ssp %d, hdmi %d", sdw_num, ssp_num, hdmi_num);
+
+	num_links = ssp_num + sdw_num + hdmi_num;
+	links = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link) *
+			     num_links, GFP_KERNEL);
+	cpus = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link_component) *
+			     num_links, GFP_KERNEL);
+	if (!links || !cpus)
+		goto devm_err;
+
+	/* SDW */
+	if (!sdw_num)
+		goto SSP;
+
+	link = mach_params->links;
+	while (link->num_adr) {
+		val = create_sdw_codec_dai(dev, links, id, cpus, link);
+		if (val < 0) {
+			dev_err(dev, "failed to create dai link %d", id);
+			goto devm_err;
+		}
+
+		id += val;
+		link++;
+	}
+
+SSP:
+	/* SSP */
+	if (ssp_num > 0) {
+		init_dai_link(links, id, "SSP2-Codec", cpus, "SSP2 Pin",
+			      rt1308_component, 1);
+
+		links[id].dpcm_playback = 1;
+		links[id].ops = &rt1308_i2s_ops,
+
+		id++;
+	}
+
+#if IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)
```

Comment:
@RanderWang  This is in old driver as well, but this won't work if only common hdmi is enabled in KCONFIG. You should check for both common hdmi and hdmi in the ifdef.

Reply:
Thanks I will check it tomorrow

Reply:
let's only support the new common HDMI and not HDAC_HDMI. This is new stuff, we don't have to be compatible with  hdac_hdmi (and btw we don't even test with hdac_hdmi any longer).